[core] Add layer block that returns colored wire

seldridge opened the pull request:
Change the `layer.block` API to return values.  If the value is a `Data`,
then create a wire of layer-colored probe type immediately before the
layer block.  Inside the layer block, define this value with a probe of
the return.  If the value is a non-`Data`, then return it directly.

Previously, this API would always return `Unit`.

Add a type class, to be used for layer blocks, that can be used to
post-process a layer block's return value.  Two type class generators are
provided: one that is an identity function and another that returns a
wire of layer-colored probe type.

This latter type class generator has the effect of creating code like the
following without requiring a pre-declared layer:

``` firrtl
wire a: Probe<UInt<1>>
layerblock A:
  wire b: UInt<1>
  define a = probe(b)
```

CC: @hcook 

This is stacked on https://github.com/chipsalliance/chisel/pull/4628.

#### Release Notes

Change `layer.block` API to allow for values to be returned. If a value is a `Data`, this will return a `Wire` of layer-colored probe type. If a non-`Data`, the value will be returned directly. Advanced users may override this behavior via a typeclass, `BlockReturnHandler`.

dtzSiFive reviewed (changes requested):
This has some problems.  The new construct isn't supported elsewhere (e.g., Select) and it's unclear what this "Region" should mean.

Marking "request changes" because it isn't suitable for use as-is, specifically the generated wire cannot be used.

seldridge commented:
Nice suggestion, Will.

I refactored, rebased, and force-pushed using an approach along the lines of your suggestion. This changes a `Block` to be a wrapper around a sequence of command buffers. The "append point" (the command buffer within a block to which commands will be written to) can then be changed with the standard thunk-based APIs for doing this safely.

jackkoenig commented on core/src/main/scala/chisel3/internal/firrtl/IR.scala:
```diff
@@ -350,17 +354,32 @@ private[chisel3] object ir {
+      _commandBufferList += _appendPoint
+      oldPoint
+    }
 
```
Discussed offline a bit that I like the direction and we could probably use this basic functionality in the future to enable certain APIs that need to declare wires outside of the current when-scope (recursively, effectively removing it from any when scoping to the top-level scope in the current module).

I do think we want to be mindful of the number and size of objects we're allocating. My hope is that we can make this "common path" of little to nothing being appended cheap and fast, while still retaining the flexibility you're adding.

I have a different idea. What if, instead of adding this complexity to Blocks, we make it such that whenever a layerblock would occur there are actually 2 Commands appended to the current commands rather than 1.
The first command is the "append point" of the previous block, the simplest implementation would be to use a Block that is empty and only contains anything if Commands are appended there. The 2nd command would then be the Block representing the Layer.

An optimization would be to turn that first command into an "EmptyCommand" that gets mutably swapped out with a Block only if appends occur. This would require making the contents of a Block an ArrayBuffer rather than an ArraySeq Builder, but that is probably a worthwhile tradeoff.

Does this make sense? Is there an obvious problem I'm missing?

seldridge replied:
> Does this make sense? Is there an obvious problem I'm missing?

This was the original approach. That `Region` that was discussed above was a command that contained a `Block` where you could write commands later.

The problem with this is that it requires rethinking how scope checking works. Currently, a block is the same as a scope. Data is only visible to you if the current block (as viewed by the builder) is in an ancestor block. However, if we introduce a sibling block that you can write to, then that breaks this.

Hence, I went with this alternative approach which keeps the notion of a block being the same as a scope, but fracturing a block internally to be multiple command buffers.

We could add another command that contains a command buffer, but is not a block and hence not a scope.

WDYT?

jackkoenig replied:
Sorry I forgot to respond earlier but that makes sense, thank you for the explanation. 

> We could add another command that contains a command buffer, but is not a block and hence not a scope.

This matches my intuition as the best way to do it, if you could give it a try and see if it works I would appreciate it. Maybe it's a bad idea then we can revert back to this approach.

seldridge commented:
This is now really work-in-progress. I've got an approach which switches to a `Placeholder` object wrapping a `Placeholder` IR object which duplicates some functionality of a `Block` (it has a command buffer), but it is _not_ a `Block` and will be subject to the same `Block` restrictions.

I need to dramatically clean up this patch to better incorporate the TypeClass approach that Jack worked out (shown in the `block2` API) which can be used to unify the `block` and `blockWithRegion` APIs into a single one. I'll clean this up tomorrow.

Thanks for all the great feedback. If you have any comments on the `Placeholder` approach, let me know. This will need to be updated to work with the Select APIs.

jackkoenig commented on core/src/main/scala/chisel3/Layer.scala:
```diff
@@ -198,6 +199,77 @@ object layer {
   }
 
+  /** A type class that describes how to post-process the return value from a layer block. */
+  trait BlockReturnHandler[A] {
+
+    /** Post-process the return value.
+      *
+      * @param placeholder the location above the layer block where commands can be inserted
+      * @param layerBlock the current layer block
+      * @param result the return value of the layer block
+      * @param sourceInfo source locator information
+      * @return the post-processed result
+      */
+    def apply(placeholder: Placeholder, layerBlock: LayerBlock, result: A)(implicit sourceInfo: SourceInfo): A
+  }
```
Related to making my suggestion to make `Placeholder` private, doing so implies that this apply method should be [package] private which then implies that this trait should be sealed. I agree with all of these implications--this is extremely powerful stuff here and I'd rather keep it private and expose various things as little pieces rather than opening it all up.

FWIW I am moving Lookupable in this direction--it was a mistake to leave it unsealed so I've had to deprecate inheritance to fix that: https://github.com/chipsalliance/chisel/blob/7ba079dc51de726cfb65b7aeb3e07ad1b9dfae1c/core/src/main/scala/chisel3/experimental/hierarchy/core/Lookupable.scala#L42. The solution for user extensibility lies in factory methods that let you implement the typeclass without having to be able to implement the virtual methods, see `Lookupable.productX` [[1]](https://github.com/chipsalliance/chisel/blob/7ba079dc51de726cfb65b7aeb3e07ad1b9dfae1c/core/src/main/scala/chisel3/experimental/hierarchy/core/Lookupable.scala#L116) and examples [[2]](https://www.chisel-lang.org/docs/cookbooks/hierarchy#how-do-i-make-case-classes-with-lots-of-fields-accessible-from-an-instance) for what I mean.

seldridge replied:
This plan sounds good to me. I did the minimal thing of making this private in the prerequisite PR and then making the trait sealed and the `apply` method package private in https://github.com/chipsalliance/chisel/pull/4623/commits/e240e7d67ceba440ff33e69123362a289471173e. This needed a rebase, but the original commit order is unchanged.

seldridge commented:
@jackkoenig and @dtzSiFive: I changed this _back_ to return `(_ <: Data) | Unit` in some fixup commits. This uses a derivative of your patch, @jackkoenig.

dtzSiFive commented on core/src/main/scala/chisel3/Layer.scala:
```diff
@@ -196,6 +197,100 @@ object layer {
+        }
+
+        // Similarly, if the result is a probe, then we forward it directly.  If
+        // it is a non-probe, then we need to probe if first.
```
```suggestion
        // it is a non-probe, then we need to probe it first.
```

seldridge replied:
Fixed in: cbc862ea1

dtzSiFive commented on core/src/main/scala/chisel3/Layer.scala:
```diff
@@ -196,6 +197,100 @@ object layer {
+          }
+        }
+
+        // Similarly, if the result is a probe, then we forward it directly.  If
```
FWIW lack of probe-ness at the top level of an aggregate doesn't mean the interior doesn't have probes; similarly if they contain properties (or probes) they can't be probed.

The simplification made here (Data is either probe or not-probe and if not-probe makes sense to be probed) is one done in at least a few other places, so don't mean to snipe unduly here.  It may be suitable in practice?

I haven't put my solution hat on for this lately, so just FYI!

I'm not sure whether Chisel will presently catch this or not.  If not it probably should when constructing the `probe<T>`.  Could be useful to intercept and diagnose the problem here?

Lots of thoughts, leaving this short.

seldridge replied:
Yeah, this is a problem. I'm going to see about handling it in a follow-on.

A related problem is being able to return layer-colored `Data`. I improved this in the rebase to work more directly. See https://github.com/chipsalliance/chisel/pull/4623/files#diff-d85d136248243112b9d589c7160462146720114db92669778a6d2831ab73d3b0R343 for "up casting" from `Probe<UInt>` to `Probe<UInt, A>`. See https://github.com/chipsalliance/chisel/pull/4623/files#diff-d85d136248243112b9d589c7160462146720114db92669778a6d2831ab73d3b0R427 for checking of returning an illegal probe type and this being caught in Chisel.

For aggregates, we will need to do some deeper inspection and potentially do things like what is happening with the up-casting here on a fine-grained basis. CIRCT will have to act for the backstop for this for now...

dtzSiFive commented on core/src/main/scala/chisel3/Layer.scala:
```diff
@@ -196,6 +197,100 @@ object layer {
+
+        // Similarly, if the result is a probe, then we forward it directly.  If
+        // it is a non-probe, then we need to probe if first.
+        Builder.forcedUserModule.withRegion(layerBlock.region) {
```
I don't think it matters given the specific hardware created here (?), but just a note that stands out: this creates hardware assuming that changing the block is sufficient (with Placeholder and related builder-state management details fresh on our minds).  I think that's right but is important this only does things where having the rest of the state wrong is not an issue (e.g., layerStack).

dtzSiFive commented on core/src/main/scala/chisel3/Layer.scala:
```diff
@@ -196,6 +197,100 @@ object layer {
+    /** Return a [[BlockReturnHandler]] that, if a layer block is created, will
+      * create a [[Wire]] of layer-colored probe type _above_ the layer block
+      * and [[probe.define]] this at the end of the layer block.  If no
+      * layer-block is created, then the result is pass-through.
```
Cool!  This has some neat properties!

dtzSiFive reviewed (approved):
LGTM!

Can this break existing code that today "returns" something from a block (I am not sure offhand if terminating with a value will error or just be dropped today (pre-PR)) that might now be a problem?

Mostly re:handling of release flow/target version details.

dtzSiFive commented:
Up to you, but if this is derived from @jackkoenig's code consider including the co-authored-by information for attribution.

seldridge commented:
> Can this break existing code that today "returns" something from a block (I am not sure offhand if terminating with a value will error or just be dropped today (pre-PR)) that might now be a problem?

Good question. Not a problem given that the return value today is always `Unit`. If a user is trying to return something today, they can't do it through the return of the layer block and have to mutably set a variable, defined above the layer block, outside the layer block. This patch is a change to existing behavior, but not in a way which should break anything (though it is possible). E.g., this will break code that is doing a runtime check that the return type is a `Unit`:

```
val foo = layer.block(A) { Wire(Bool()) }
assert(foo.getClass() == classOf[Unit])
```

jackkoenig reviewed (approved):
LGTM! Nice work!

mwachs5 commented on src/test/scala/chiselTests/LayerSpec.scala:
```diff
@@ -306,6 +306,107 @@ class LayerSpec extends ChiselFlatSpec with Utils with FileCheck {
+      }
+
+      // A single layer block
+      val a = layer.block(A) {
```
super cool! 

One thing, is it important to explain why this must be a read-only probe vs read-write? Is that a design choice or a requirement ? Is there potential future work like `layer.writableblock(A)` or is that fundamentally impossible or a "we dont need it yet so let's not do it yet"?

seldridge replied:
We're not sure, yet, if we want to allow writes to layers. This patch is close enough to allowing this if we want to go that way, though. To do that, you would probably need to manually return a `RWProbe` (and this would need to be updated to keep the write flag). Something like:

``` scala
val a = layer.block(A) {
  Wire(probe.RWProbe(Bool()))
}
```

(This patch will let you do this, but you will get a read probe wire as the return. If you manually make this a RWProbe and try to drive it, CIRCT will have problems.)

Towards the problem of having data go up and down, I'm going to first see how far we can push the `Placeholder` API to allow for construction of hardware in the parent while constructing the child.

dtzSiFive commented:
> ```scala
> val foo = layer.block(A) { Wire(Bool()) }
> assert(foo.getClass() == classOf[Unit])
> ```

If the returned Data is not safe to create a probe out of, no `assert` needed.
I doubt this breaks anything currently out there, similarly to "but what if you checked the return type".

seldridge commented:
> If the returned Data is not safe to create a probe out of, no assert needed.

Yes, this is one place where this could break things. If the last statement is a `Data`, but is not legal to create a `Probe` type out of, this will cause a Chisel elaboration-time failure.
